firedrake/function.py

```diff
@@ -354,13 +355,22 @@ def _interpolate(self, fs, dat, expression, subset):
             raise RuntimeError('Rank mismatch: Expression rank %d, FunctionSpace rank %d'
                                % (expression.rank(), len(fs.ufl_element().value_shape())))
 
-        if expression.value_shape() != fs.ufl_element().value_shape():
+        if expression.ufl_shape != fs.ufl_element().value_shape():
             raise RuntimeError('Shape mismatch: Expression shape %r, FunctionSpace shape %r'
-                               % (expression.value_shape(), fs.ufl_element().value_shape()))
+                               % (expression.ufl_shape, fs.ufl_element().value_shape()))
 
         coords = fs.mesh().coordinates
 
-        if hasattr(expression, "eval"):
+        if not isinstance(expression, expression_t.Expression):
+            kernel, oriented, coefficients = self._interpolate_ufl_expression(expression, to_pts, to_element, fs)
+            args = [kernel, subset or self.cell_set,
+                    dat(op2.WRITE, fs.cell_node_map()[op2.i[0]])]
```

Comment:
the `op2.i` part indicates that this kernel argument is a "buffer", (like the normal output tensor buffer).  What is the failure mode?  I haven't looked at what the kernel you produce looks like for the VectorFunctionSpace case (and whether that matches up correctly).